### PR TITLE
Avoid additional errors once the appium session has crashed

### PR DESCRIPTION
**Pull request by Cawllec**
## Goal

In the case that the appium session has fully crashed, we need to stop attempting to invoke appium methods, as doing so will always spawn an error that's noisy:
- In the test itself, as it'll be more immediate when viewing the test logs
- In the maze-runner dashboard, as these will occur fairly frequently if the appium session crashes

This change seeks to remove these additional errors by:
- Adding an error captor class will record each error occurring in a scenario, being reset at the end of said scenario
- After each scenario involving appium that fails, the captor class will be consulted to see if the appium session crashed during the scenario
- If the appium session did crash, the driver will be `nil`-ed
- Subsequent methods will check the driver isn't `nil` before they attempt to invoke it again

## Tests

Unit tests have been added for the error captor, this update will be included in an overnight build of android and cocoa to see if we can see full process working in the case of an appium error.

### CODE REVIEW

**Comment by Cawllec** [comment]
Something has gone wrong here - I'll let you know when it's ready for review @twometresteve

File: lib/maze/bugsnag_config.rb [hunk 1]
```diff
@@ -88,5 +88,18 @@ def call(report)
         @middleware.call(report)
       end
     end
+
+    class ErrorCaptureMiddleware
```
Comment (twometresteve) [inline lib/maze/bugsnag_config.rb:92]: This this needs to be `use`d after line 35?

**Review by twometresteve — COMMENTED** [review]
Looking very promising - my comments are pretty trivial I think.

File: lib/maze/hooks/appium_hooks.rb [hunk 2]
```diff
@@ -73,6 +84,15 @@ def at_exit
 
       private
 
+      def driver_has_failed?
+        error_captor = Maze::ErrorCaptor.instance
+        return false unless error_captor.captured_errors_exist?
+
+        error_captor.get_captured_error_classes.any? do |error_class|
+          APPIUM_DRIVER_FAILED_ERRORS.include?(error_class)
+        end
```
[the remark this answers is not in the document]
Comment (Cawllec) [reply lib/maze/hooks/appium_hooks.rb:93]: We still need it in this case, as `[].any?` is false.

**Review by twometresteve — APPROVED** [review]

**Pull request merged**